pin/gc/triset.go

```diff
@@ -0,0 +1,160 @@
+package gc
+
+import (
+	"context"
+
+	dag "github.com/ipfs/go-ipfs/merkledag"
+
+	cid "gx/ipfs/QmTprEaAA2A9bst5XH7exuyi5KzNMK3SEDNN8rBDnKWcUS/go-cid"
+)
+
+const (
+	colorNull trielement = iota
+	color1
+	color2
+	color3
+	colorMask = 0x3
+)
+
+const (
+	// enum from enumerator
+	// stricter enumerators should have higer value than those less strict
+	enumFast   trielement = 0
+	enumStrict trielement = 1 << 2
+	enumMask   trielement = enumStrict
+)
+
+// this const is used to enable runtime check for things that should never happen
+// will cause panic if they happen
+const pedantic = true
+
+// try to keep trielement as small as possible
+// 8bit color is overkill, 3 bits would be enough, but additional functionality
+// will probably increase it future
+// if ever it blows over 64bits total size change the colmap in triset to use
+// pointers onto this structure
+type trielement uint8
+
+func (t trielement) getColor() trielement {
+	return t & colorMask
+}
+
+func (t trielement) getEnum() trielement {
+	return t & enumMask
+}
+
+type triset struct {
+	// colors are per triset allowing fast color swap after sweep,
+	// the update operation in map of structs is about 3x as fast
+	// as insert and requres 0 allocations (keysize allocation in case of insert)
+	white, gray, black trielement
+
+	freshColor trielement
+
+	// grays is used as stack to enumerate elements that are still gray
+	grays []*cid.Cid
+
+	// if item doesn't exist in the colmap it is treated as white
+	colmap map[string]trielement
+}
+
+func newTriset() *triset {
+	tr := &triset{
+		white: color1,
+		gray:  color2,
+		black: color3,
+
+		grays:  make([]*cid.Cid, 0, 1<<10),
+		colmap: make(map[string]trielement),
+	}
+
+	tr.freshColor = tr.white
+	return tr
+}
+
+// InsertFresh inserts fresh item into a set
+// it marks it with freshColor if it is currently white
+func (tr *triset) InsertFresh(c *cid.Cid) {
+	e := tr.colmap[c.KeyString()]
```

Comment:
`e, ok := ....`, right? i guess its just a uint8, and 0 is colorNull?

Reply:
yes, 0 is null value